[SPMD] Make compile cache the compilation result and add option to perform transformation

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97836

This PR changes ``compile()`` decorator to cache the compilation result so that the compilation is done once. An gm_transformation option is also added to ``compile()`` so that after the compilation is done, users can perform any graph transformation with the compiled graph module.

Differential Revision: [D44484033](https://our.internmc.facebook.com/intern/diff/D44484033/)

fegin committed:
[SPMD] Make compile to cache the compilation result and add option to perform transformation

This PR changes ``compile()`` decorator to cache the compilation result so that the compilation is done once. An gm_transformation option is also added to ``compile()`` so that after the compilation is done, users can perform any graph transformation with the compiled graph module.

Differential Revision: [D44484033](https://our.internmc.facebook.com/intern/diff/D44484033/)

[ghstack-poisoned]

wconstab commented:
what's the reason for wanting a @compile wrapper and a cache mechanism for the compiled obj?

would it be easier if you just used a different compile method such as `compiled_func = compile(func)` so there is no confusion about which obj you are calling at runtime?

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -233,102 +344,38 @@ def compile(module_override: Optional[Dict[Type[Any], Override]] = None):
+                    compiled_obj.params_and_buffers
+                )[0]
+                if first_iter and gm_transformation:
+                    # TODO: SPMD should provid a default and configurable
```
+1 to this

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -219,8 +221,117 @@ def f_true():
+    return _CompiledResult(gm, mod, opt, named_states, params_and_buffers)
+
+
+COMPILED_OBJECT_KEY = "_compiled_obj"
```
this can simply be an `COMPILED_OBJECT_KEY =object()` to avoid potential string name conflicts?

fegin replied:
I remember there was a serialization issue due to using `object()` for composable APIs. But I guess this is not a problem unless we want to serialize the `train_step`. Will change it to `object()`

fegin replied:
Hmm, I just remember that the convention of `__dict__` is that keys must be strings. So I will keep str but add a TODO to ensure the key is unique.

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -219,8 +221,117 @@ def f_true():
+    params_and_buffers: List[torch.Tensor]
+
+
+def _compile(
```
assuming no code change in this function.

fegin replied:
Nothing change except that the extra return object `_CompiledResult`.

mrshenli reviewed (approved):
LGTM

wconstab reviewed (approved):
changes in this PR lgtm.  My other comment is more about the existing @compile wrapper, i guess.  I think the wrapper looks like it might do advanced things like caching/recompiling but it actually doesn't.  I guess my suggestion would be just use an explicit compile step so its simpler.  But we can ignore that and leave it as is.

mrshenli commented:
Lint error is real:

```
>>> Lint for torch/distributed/_spmd/api.py:

  Error (MYPY) [arg-type]
    Argument 3 to "_CompiledResult" has incompatible type
    "Optional[Optimizer]"; expected "Optimizer"

        322  |        for _, override in module_override.items():
        323  |            gm = override.transform(gm, name_to_spec)
        324  |
    >>> 325  |    return _CompiledResult(gm, mod, opt, named_states, params_and_buffers)
        326  |
        327  |
        328  |COMPILED_OBJECT_KEY = "_compiled_obj"

  Error (MYPY) [arg-type]
    Argument 5 to "_CompiledResult" has incompatible type "Dict[str,
    Union[Tensor, Parameter]]"; expected "List[Tensor]"

        322  |        for _, override in module_override.items():
        323  |            gm = override.transform(gm, name_to_spec)
        324  |
    >>> 325  |    return _CompiledResult(gm, mod, opt, named_states, params_and_buffers)
        326  |
        327  |
        328  |COMPILED_OBJECT_KEY = "_compiled_obj"
```

fegin committed:
Update on "[SPMD] Make compile cache the compilation result and add option to perform transformation"


This PR changes ``compile()`` decorator to cache the compilation result so that the compilation is done once. An gm_transformation option is also added to ``compile()`` so that after the compilation is done, users can perform any graph transformation with the compiled graph module.

Differential Revision: [D44484033](https://our.internmc.facebook.com/intern/diff/D44484033/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Make compile cache the compilation result and add option to perform transformation"


This PR changes ``compile()`` decorator to cache the compilation result so that the compilation is done once. An gm_transformation option is also added to ``compile()`` so that after the compilation is done, users can perform any graph transformation with the compiled graph module.

Differential Revision: [D44484033](https://our.internmc.facebook.com/intern/diff/D44484033/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Make compile cache the compilation result and add option to perform transformation"


This PR changes ``compile()`` decorator to cache the compilation result so that the compilation is done once. An gm_transformation option is also added to ``compile()`` so that after the compilation is done, users can perform any graph transformation with the compiled graph module.

Differential Revision: [D44484033](https://our.internmc.facebook.com/intern/diff/D44484033/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[SPMD] Make compile cache the compilation result and add option to perform transformation"


This PR changes ``compile()`` decorator to cache the compilation result so that the compilation is done once. An gm_transformation option is also added to ``compile()`` so that after the compilation is done, users can perform any graph transformation with the compiled graph module.

Differential Revision: [D44484033](https://our.internmc.facebook.com/intern/diff/D44484033/)

[ghstack-poisoned]

fegin commented:
The following error is not related.

[linux-bionic-py3_8-clang8-xla / test (xla, 1, 1, linux.4xlarge)](https://github.com/pytorch/pytorch/actions/runs/4555841736/jobs/8035902031?pr=97836#logs)

Error: Can't find 'action.yml', 'action.yaml' or 'Dockerfile' under '/home/ec2-user/actions-runner/_work/pytorch/pytorch/.github/actions/get-workflow-job-id'. Did you forget to run actions/checkout before running your local action?

fegin commented:
@pytorchbot merge
